_docs/rfc/for_each_iteration.md

```diff
@@ -0,0 +1,218 @@
+  order is now adjusted as:
+    - `for_each` and expand loop.
+    - `locals`
+    - `include`
+    - `dependency`
+    - Everything else
+- Support `merge_parent` in `remote_state`, which ultimately adjusts how `mergeConfigWithIncludedConfig` merges the
+  child `remote_state` config with the parent.
+
+
+##### Pros and Cons of approach
+
+**Pros**:
+
+- Parallels `for_each` in Terraform. In the examples I used `list`, but the implementation details can mimic that of
+  Terraform (requiring `set` or `map`). That way, the syntax closely resembles an existing construct and thus it will be
+  easier to pick up.
+- Limited changes are necessary in the user config to add in iteration.
+- The config parser implementation is relatively simple.
+
+**Cons**:
+
+- The implementation of the runners can become complex, especially in the dependency detection for the `xxx-all`
+  commands. If `dependency` blocks are parameterized by looping, this can be hairy to unpack and expand into the
+  dependency graph if the terragrunt files don't exist.
+- The logic for expanding out the loop for the terragrunt cache is complex as it is not immediately obvious what the
+  folder structure should be. Should it be a hash of the iteration value appended to the working path (e.g
+  `.terragrunt-cache/HASH_OF_EACH_VALUE/path/to/workspace`)? Or should it be one cache per iteration value (e.g
+  `HASH_OF_EACH_VALUE/.terragrunt-cache/path/to/workspace`)?
+- Implementation requires a major refactor in the calling logic post parsing.
+- It is not immediately obvious how to handle removing an item in the `for_each` list. A naive approach would not run
+  `terragrunt destroy` on that module. Additionally, as stated, it is impossible to run the command on a single item in
+  the list and thus you can't destroy just the items you want to remove without doing complex dancing (e.g update the
+  `for_each` list to only reference the items you want to destroy, then run `terragrunt destroy` to destroy just those
+  items, and then restore the list to the version with the items removed).
+
+### Option 3: scaffolding tool that code gens live config using a template
+
+Instead of implementing looping within the terragrunt execution logic, this approach proposes generating the live config
+through templating. The idea here would be to use a templating tool (e.g
+[cookiecutter](https://github.com/cookiecutter/cookiecutter)) to render each live config parameterized by some list
+value that will then generate the folder structure.
+
+In this approach, no change is necessary within Terragrunt to support this other than perhaps documentation and examples
+to demonstrate the approach.
+
+**Pros**:
+
+- Keeps Terragrunt clean, since no new code is introduced.
+- Available workaround now, without any work.
+
+**Cons**:
+
+- Code generation is complex. When should you run the code generation step? How often? How do you know if there is
```

Comment:
Do you check in the generated code? If so, you have to maintain it (e.g., 19 sets of files/folders for 19 AWS regions). If it's always generated at runtime, then interacting with just one of the generated things becomes harder.

Reply:
Valid points! Adding this in as one of the concerns.